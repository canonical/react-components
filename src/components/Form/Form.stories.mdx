import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Form from "./Form";
import Input from "../Input";
import Select from "../Select";

<Meta
  title="Form"
  component={Form}
  argTypes={{
    children: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => <Form {...args} />;

### Form

This is a [React](https://reactjs.org/) component for the Vanilla [Form](https://docs.vanillaframework.io/base/forms).

Form controls have global styling defined at the HTML element level. Labels and most input types are 100% width of the `<form>` parent element.

### Props

<ArgsTable of={Form} />

### Forms

<Canvas>
  <Story
    name="Default"
    args={{
      children: (
        <>
          <Input
            type="text"
            id="exampleInputEmail12"
            placeholder="example@canonical.com"
            label="Email address"
          />
          <Input
            type="password"
            id="exampleInputPassword12"
            placeholder="******"
            label="Password"
          />
          <Input type="file" id="exampleInputFile2" label="File input" />
          <Input
            type="checkbox"
            id="CheckMe2"
            label="I agree to receive information about Canonicalâ€™s products and services."
          />
        </>
      ),
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Fieldset

<Canvas>
  <Story name="Fieldset">
    <Form>
      <fieldset>
        <Input
          placeholder="Joe"
          id="list-input-12"
          type="text"
          label="First name"
        />
        <Input
          placeholder="Bloggs"
          id="list-input-22"
          type="text"
          label="Last name"
        />
        <Input
          placeholder="example@canonical.com"
          id="list-input-32"
          type="text"
          label="Email address"
        />
      </fieldset>
    </Form>
  </Story>
</Canvas>

### Inline

<Canvas>
  <Story name="Inline">
    <Form inline>
      <Input
        type="text"
        id="username-inline2"
        className="p-form__control"
        label="Username"
        help="30 characters or fewer."
      />
      <Input
        type="text"
        id="address-inline22"
        aria-invalid="true"
        label="Email address"
        error="Please enter a valid email address."
      />
    </Form>
  </Story>
</Canvas>

### Stacked

<Canvas>
  <Story name="Stacked">
    <Form stacked>
      <Input type="text" id="full-name-stacked2" label="Full name" stacked />
      <Input
        type="text"
        id="username-stacked2"
        label="Username"
        stacked
        help="30 characters or fewer."
      />
      <Input
        type="text"
        id="username-stacked-error2"
        label="Email address"
        stacked
        error="This field is required"
      />
      <Input
        type="text"
        id="address-optional-stacked2"
        label="Address line 1"
        stacked
      />
      <Input
        type="text"
        id="address-optional-stacked3"
        label="Address line 2"
        stacked
      />
    </Form>
  </Story>
</Canvas>

### Disabled

<Canvas>
  <Story name="Disabled">
    <Form>
      <Input
        label="Email address"
        type="text"
        id="disabled-input2"
        placeholder="example@canonical.com"
        disabled="disabled"
      />
    </Form>
  </Story>
</Canvas>

### Validation

<Canvas>
  <Story name="Validation">
    <Form>
      <Input
        type="text"
        id="exampleTextInputError2"
        placeholder="example@canonical.com"
        label="Email address"
        error="This field is required."
      />
      <Input
        type="text"
        id="exampleTextInputCaution2"
        placeholder="14"
        label="Mail configuration ID"
        caution="No validation is performed in preview mode."
      />
      <Input
        type="text"
        id="exampleTextInputSuccess2"
        placeholder="**** **** **** ****"
        label="Card number"
        success="Verified."
      />
      <Select
        id="exampleSelectInputError3"
        defaultValue=""
        options={[
          { value: "", disabled: "disabled", label: "--Select an option--" },
          { value: "1", label: "Cosmic Cuttlefish" },
          { value: "2", label: "Bionic Beaver" },
          { value: "3", label: "Xenial Xerus" },
        ]}
        label="Ubuntu releases"
        error="You need to select an OS to complete your install."
      />
    </Form>
  </Story>
</Canvas>

### Required

<Canvas>
  <Story name="Required">
    <Form>
      <Input
        type="text"
        id="exampleTextInputError3"
        placeholder="e.g joe@bloggs.com"
        error="This field is required."
        label="Email address"
        required
      />
    </Form>
  </Story>
</Canvas>
