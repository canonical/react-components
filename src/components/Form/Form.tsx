import classNames from "classnames";
import React from "react";
import type { HTMLProps, ReactNode } from "react";

import type { ClassName, PropsWithSpread } from "types";

export type Props = PropsWithSpread<
  {
    /**
     * The content of the form.
     */
    children?: ReactNode;
    /**
     * Optional class(es) to pass to the form element.
     */
    className?: ClassName;
    inline?: boolean;
    stacked?: boolean;
  },
  HTMLProps<HTMLFormElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Form](https://docs.vanillaframework.io/base/forms).
 *
 * Form controls have global styling defined at the HTML element level. Labels and most input types are 100% width of the `<form>` parent element.
 */
const Form = ({
  children,
  className,
  inline,
  stacked,
  ...props
}: Props): JSX.Element => (
  <form
    className={classNames(className, {
      "p-form": inline || stacked,
      "p-form--inline": inline,
      "p-form--stacked": stacked,
    })}
    {...props}
  >
    {children}
  </form>
);

export default Form;
