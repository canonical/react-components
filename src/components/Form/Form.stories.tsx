import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Form from "./Form";
import Input from "../Input";
import Select from "../Select";

const meta: Meta<typeof Form> = {
  component: Form,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Form>;

export const Default: Story = {
  name: "Default",

  args: {
    children: (
      <>
        <Input
          type="text"
          id="exampleInputEmail12"
          placeholder="example@canonical.com"
          label="Email address"
        />
        <Input
          type="password"
          id="exampleInputPassword12"
          placeholder="******"
          label="Password"
        />
        <Input type="file" id="exampleInputFile2" label="File input" />
        <Input
          type="checkbox"
          id="CheckMe2"
          label={
            "I agree to receive information about Canonicalâ€™s products and services."
          }
        />
      </>
    ),
  },
};

export const Fieldset: Story = {
  render: () => (
    <Form>
      <fieldset>
        <Input
          placeholder="Joe"
          id="list-input-12"
          type="text"
          label="First name"
        />
        <Input
          placeholder="Bloggs"
          id="list-input-22"
          type="text"
          label="Last name"
        />
        <Input
          placeholder="example@canonical.com"
          id="list-input-32"
          type="text"
          label="Email address"
        />
      </fieldset>
    </Form>
  ),

  name: "Fieldset",
};

export const Inline: Story = {
  render: () => (
    <Form inline>
      <Input
        type="text"
        id="username-inline2"
        className="p-form__control"
        label="Username"
        help="30 characters or fewer."
      />
      <Input
        type="text"
        id="address-inline22"
        aria-invalid="true"
        label="Email address"
        error="Please enter a valid email address."
      />
    </Form>
  ),

  name: "Inline",
};

export const Stacked: Story = {
  render: () => (
    <Form stacked>
      <Input type="text" id="full-name-stacked2" label="Full name" stacked />
      <Input
        type="text"
        id="username-stacked2"
        label="Username"
        stacked
        help="30 characters or fewer."
      />
      <Input
        type="text"
        id="username-stacked-error2"
        label="Email address"
        stacked
        error="This field is required"
      />
      <Input
        type="text"
        id="address-optional-stacked2"
        label="Address line 1"
        stacked
      />
      <Input
        type="text"
        id="address-optional-stacked3"
        label="Address line 2"
        stacked
      />
    </Form>
  ),

  name: "Stacked",
};

export const Disabled: Story = {
  render: () => (
    <Form>
      <Input
        label="Email address"
        type="text"
        id="disabled-input2"
        placeholder="example@canonical.com"
        disabled
      />
    </Form>
  ),

  name: "Disabled",
};

export const Validation: Story = {
  render: () => (
    <Form>
      <Input
        type="text"
        id="exampleTextInputError2"
        placeholder="example@canonical.com"
        label="Email address"
        error="This field is required."
      />
      <Input
        type="text"
        id="exampleTextInputCaution2"
        placeholder="14"
        label="Mail configuration ID"
        caution="No validation is performed in preview mode."
      />
      <Input
        type="text"
        id="exampleTextInputSuccess2"
        placeholder="**** **** **** ****"
        label="Card number"
        success="Verified."
      />
      <Select
        id="exampleSelectInputError3"
        defaultValue=""
        options={[
          {
            value: "",
            disabled: true,
            label: "--Select an option--",
          },
          {
            value: "1",
            label: "Cosmic Cuttlefish",
          },
          {
            value: "2",
            label: "Bionic Beaver",
          },
          {
            value: "3",
            label: "Xenial Xerus",
          },
        ]}
        label="Ubuntu releases"
        error="You need to select an OS to complete your install."
      />
    </Form>
  ),

  name: "Validation",
};

export const Required: Story = {
  render: () => (
    <Form>
      <Input
        type="text"
        id="exampleTextInputError3"
        placeholder="e.g joe@bloggs.com"
        error="This field is required."
        label="Email address"
        required
      />
    </Form>
  ),

  name: "Required",
};
