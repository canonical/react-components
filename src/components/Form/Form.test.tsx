import { render, screen } from "@testing-library/react";
import React from "react";

import Form from "./Form";

describe("Form ", () => {
  it("renders", () => {
    render(
      <Form name="this needs a name so RTL can find it">Test content</Form>
    );
    expect(screen.getByRole("form")).toMatchSnapshot();
  });

  it("can be inline", () => {
    render(<Form name="this needs a name so RTL can find it" inline={true} />);
    expect(screen.getByRole("form")).toHaveClass("p-form--inline");
  });

  it("can be stacked", () => {
    render(<Form name="this needs a name so RTL can find it" stacked={true} />);
    expect(screen.getByRole("form")).toHaveClass("p-form--stacked");
  });

  it("can add additional classes", () => {
    render(
      <Form
        name="this needs a name so RTL can find it"
        stacked={true}
        className="extra-class"
      />
    );
    const form = screen.getByRole("form");
    expect(form).toHaveClass("p-form");
    expect(form).toHaveClass("extra-class");
  });
});
