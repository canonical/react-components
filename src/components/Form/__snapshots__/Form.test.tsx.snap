// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form  renders 1`] = `
<form
  class=""
  name="this needs a name so RTL can find it"
>
  Test content
</form>
`;
